### PR TITLE
Enable sriov

### PR DESCRIPTION
**Pull request by Boredphilosopher96**

### CODE REVIEW

File: vhpc_toolkit/command.py [hunk 1]
```diff
@@ -65,3 +65,5 @@ def command():
         ops.vm_scheduling_affinity_cli()
     elif ops.cfg[CMD_KEY] == "numa_affinity":
         ops.numa_affinity_cli()
+    elif ops.cfg[CMD_KEY] == "modify_sriov":
```
Comment (Michael870) [inline vhpc_toolkit/command.py]: Let's change the command to sriov_host

File: vhpc_toolkit/get_args.py [hunk 2]
```diff
@@ -739,6 +739,55 @@ def affinity_array(string):
         type=str,
         help="Name of physical function which backs up SR-IOV Passthrough",
     )
+
+    modify_sriov_parser = subparsers.add_parser(
+        "modify_sriov",
+        help="Modify SR-IOV properties on host(s).\n"
+        "This operation assumes that SR-IOV drivers have been installed on Esxi host",
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py]: ESXi

File: vhpc_toolkit/get_args.py [hunk 3]
```diff
@@ -739,6 +739,55 @@ def affinity_array(string):
...
+    modify_sriov_group1.add_argument(
+        "--file",
+        action="store",
+        default=None,
+        type=str,
+        help="Name of the file containing a list of hosts, "
+        "one per line, to perform the modify SR-IOV operation",
+    )
+    modify_sriov_parser.add_argument(
+        "--location",
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py]: Change the parameter to --device

File: vhpc_toolkit/get_args.py [hunk 4]
```diff
@@ -739,6 +739,55 @@ def affinity_array(string):
...
+        action="store",
+        default=None,
+        type=str,
+        help="Location of the pnic in format xxxx:xx:xx.x",
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py]: Location of the pnic -> Address of the device

File: vhpc_toolkit/get_args.py [hunk 5]
```diff
@@ -739,6 +739,55 @@ def affinity_array(string):
         type=str,
         help="Name of physical function which backs up SR-IOV Passthrough",
     )
+
+    modify_sriov_parser = subparsers.add_parser(
+        "modify_sriov",
+        help="Modify SR-IOV properties on host(s).\n"
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py]: properties -> configuration

File: vhpc_toolkit/get_args.py [hunk 6]
```diff
@@ -739,6 +739,55 @@ def affinity_array(string):
...
+    modify_sriov_parser.add_argument(
+        "--location",
+        action="store",
+        default=None,
+        type=str,
+        help="Location of the pnic in format xxxx:xx:xx.x",
+        required=True,
+    )
+    modify_sriov_group2.add_argument(
+        "--on", action="store_true", help="Turn on SR-IOV mode for pnic on host(s)"
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py]: Device is not necessarily a pnic. Change "pnic" to "device"

File: vhpc_toolkit/get_args.py [hunk 7]
```diff
@@ -739,6 +739,55 @@ def affinity_array(string):
...
+    )
+    modify_sriov_group2.add_argument(
+        "--off", action="store_true", help="Turn off SR-IOV mode for pnic on host(s)"
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py]: same as above

File: vhpc_toolkit/get_args.py [hunk 8]
```diff
@@ -739,6 +739,55 @@ def affinity_array(string):
...
+    )
+    modify_sriov_parser.add_argument(
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py]: Add a default value of 4

File: vhpc_toolkit/config_objs.py [hunk 9]
```diff
@@ -974,6 +974,44 @@ def create_svs(self, svs_name, vmnic, num_ports=8):
...
         host_network_obj.AddVirtualSwitch(vswitchName=svs_name, spec=svs)
 
+    def modify_sriov(
+        self,
+        pnic_location: str,
+        num_virtual_functions: int = None,
+        enable_sriov: bool = True,
+    ):
+        """
+        Function to enable/disable SRIOV and/or change the number of virtual functions on vmnics of a host.
```
Comment (Michael870) [inline vhpc_toolkit/config_objs.py]: vmnics -> devices

File: vhpc_toolkit/config_objs.py [hunk 10]
```diff
@@ -974,6 +974,44 @@ def create_svs(self, svs_name, vmnic, num_ports=8):
         host_network_obj = self.host_obj.configManager.networkSystem
         host_network_obj.AddVirtualSwitch(vswitchName=svs_name, spec=svs)
 
+    def modify_sriov(
+        self,
+        pnic_location: str,
```
Comment (Michael870) [inline vhpc_toolkit/config_objs.py]: change this variable name to "device" or "device_id"

**Review by Michael870 — CHANGES_REQUESTED** [review]
Most of the comments are on naming. Specifically, SRIOV is not limited to network, so let's change to device to be general. Also, although you see "location" used in some places, it's not common in HPC.

File: vhpc_toolkit/get_args.py [hunk 11]
```diff
@@ -739,6 +739,52 @@ def affinity_array(string):
...
+    sriov_host_group2.add_argument(
+        "--on", action="store_true", help="Turn on SR-IOV mode for device on host(s)"
+    )
+    sriov_host_group2.add_argument(
+        "--off", action="store_true", help="Turn off SR-IOV mode for device on host(s)"
+    )
+    sriov_host_parser.add_argument(
+        "--num_func",
+        action="store",
+        default=4,
```
Comment (qoofyk) [inline vhpc_toolkit/get_args.py]: remove default=4

File: vhpc_toolkit/get_args.py [hunk 12]
```diff
@@ -739,6 +739,52 @@ def affinity_array(string):
...
+        type=int,
+        help="Number of virtual functions. This argument is ignored if used with --off flag",
```
Comment (qoofyk) [inline vhpc_toolkit/get_args.py]: A warning message to users "num_func must be equal or smaller than the VF enabled in firmware"

File: vhpc_toolkit/get_args.py (hunk 8 above)
Comment (Michael870) [reply vhpc_toolkit/get_args.py]: The default should be to not change the value if user ignores it

File: vhpc_toolkit/operations.py [hunk 13]
```diff
@@ -2186,3 +2186,22 @@ def numa_affinity_cli(self):
...
+            hosts.extend(
+                [
+                    host_cfg["host"]
+                    for host_cfg in self._extract_file(self.cfg, file_keys=["host"])
+                ]
+            )
+
+        for host in hosts:
+            ConfigHost(self.objs.get_host(host)).modify_sriov(
+                self.cfg["location"],
```
Comment (Michael870) [inline vhpc_toolkit/operations.py]: location -> device

File: vhpc_toolkit/config_objs.py [hunk 14]
```diff
@@ -974,6 +974,40 @@ def create_svs(self, svs_name, vmnic, num_ports=8):
...
+        config.sriovEnabled = enable_sriov
+        if enable_sriov and num_virtual_functions:
+            config.numVirtualFunction = num_virtual_functions
+        config.id = device_id
+        try:
+            self.host_obj.configManager.pciPassthruSystem.UpdatePassthruConfig(
+                config=[config]
+            )
+            self.logger.info(
+                f"{'enabled' if enable_sriov else 'disabled'} SRIOV for network with location: {device_id} on host {self.host_obj.name}"
```
Comment (Michael870) [inline vhpc_toolkit/config_objs.py]: network with location -> device

Comment (qoofyk) [inline vhpc_toolkit/config_objs.py:1004]: SRIOV for network with location --> SRIOV on network adapter with PCIe Device ID

**Review by Michael870 — CHANGES_REQUESTED** [review]

File: vhpc_toolkit/config_objs.py [hunk 15]
```diff
@@ -974,6 +974,40 @@ def create_svs(self, svs_name, vmnic, num_ports=8):
...
+        self,
+        device_id: str,
+        num_virtual_functions: int = None,
+        enable_sriov: bool = True,
+    ):
+        """
+        Function to enable/disable SRIOV and/or change the number of virtual functions on devices of a host.
+        Number of virtual functions argument is skipped if user is trying to disable SRIOV
+        Args:
+            device_id: Address of the device which supports SRIOV in format xxxx:xx:xx.x
```
Comment (qoofyk) [inline vhpc_toolkit/config_objs.py]: device_id: PCIe address of the Virtual Function (VF) of the SRIOV device in format xxxx:xx:xx.x

File: vhpc_toolkit/config_objs.py [hunk 16]
```diff
@@ -974,6 +974,40 @@ def create_svs(self, svs_name, vmnic, num_ports=8):
...
+            num_virtual_functions: Number of virtual functions to set for the NIC
```
Comment (qoofyk) [inline vhpc_toolkit/config_objs.py]: num_virtual_functions: Number of virtual functions to set for the SRIOV device

File: vhpc_toolkit/get_args.py [hunk 17]
```diff
@@ -739,6 +739,53 @@ def affinity_array(string):
...
+        type=str,
+        help="Name of the file containing a list of hosts, "
+        "one per line, to perform the modify SR-IOV operation",
+    )
+    sriov_host_parser.add_argument(
+        "--device",
+        action="store",
+        default=None,
+        type=str,
+        help="Address of the SR-IOV device in format xxxx:xx:xx.x",
```
Comment (qoofyk) [inline vhpc_toolkit/get_args.py]: help="PCIe address of the Virtual Function (VF) of the SR-IOV device in format xxxx:xx:xx.x"

File: vhpc_toolkit/config_objs.py (hunk 14 above)
Comment (qoofyk) [reply vhpc_toolkit/config_objs.py]: network with location --> PCIe device

File: vhpc_toolkit/get_args.py (hunk 11 above)
Comment (qoofyk) [reply vhpc_toolkit/get_args.py]: Delete this comment

**Review by qoofyk — APPROVED** [review]

File: vhpc_toolkit/config_objs.py (hunk 14 above)
Comment (Michael870) [reply vhpc_toolkit/config_objs.py]: Sumukh, please double check my comment above. The request is replace three words

**Review by Michael870 — APPROVED** [review]

**Pull request merged**